### PR TITLE
[11.0] [MIG] purchase_request: Migration to 11.0 merge with purchase_request_procurement and purchase_request_to_rfq

### PR DESCRIPTION
**Pull request by HviorForgeFlow**
This module has been migrated from three different modules in order to keep functionality together. Since procurement has change in v11 seems to make sense.
Modules migrated:
- purchase_request
- purchase_request_procurement
- purchase_request_to_rfq

### CODE REVIEW

**Comment by HviorForgeFlow** [comment]
@jbeficent MTO routes are getting lost. Any idea/advice?

**Review by JordiBForgeFlow — APPROVED** [review]

**Comment by GSLabIt** [comment]
any eta to merge this one? thx

**Review by HaraldPanten — APPROVED** [review]
Functionally is working fine. Isn't it ready to approve?  THX! :+1:

File: purchase_request/views/purchase_request_view.xml [hunk 1]
```diff
@@ -0,0 +1,589 @@
...
+                    <div class="oe_edit_only">
+                        <label for="name" class="oe_inline"/>
+                        <label for="origin" class="oe_inline"/>
+                    </div>
+                    <h1>
+                        <field name="is_editable" invisible="1"/>
+                        <field name="name" class="oe_inline"
+                               attrs="{'readonly': [('is_editable','=', False)]}"/>
+                        <label string=","/>
+                        <field name="origin" class="oe_inline"
```
Comment (LoisRForgeFlow) [inline purchase_request/views/purchase_request_view.xml:68]: I would move origin to a different place, it seem a little bit weird here for me.

File: purchase_request/views/purchase_request_view.xml [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+                    <group>
+                        <group>
+                            <field name="requested_by"
+                                   attrs="{'readonly': [('is_editable','=', False)]}"/>
+                            <field name="assigned_to"
+                                   attrs="{'readonly': [('is_editable','=', False)]}"/>
+                        </group>
+                        <group>
+                            <field name="description"
+                                   attrs="{'readonly': [('is_editable','=', False)]}"/>
```
Comment (LoisRForgeFlow) [inline purchase_request/views/purchase_request_view.xml]: I reckon that `origin` field would fit fine before or after description.

File: purchase_request/views/purchase_request_view.xml [hunk 3]
```diff
@@ -0,0 +1,589 @@
...
+        <record id="purchase_request_line_form" model="ir.ui.view">
+            <field name="name">purchase.request.line.form</field>
+            <field name="model">purchase.request.line</field>
+            <field name="priority" eval="20"/>
+            <field name="arch" type="xml">
+                <form string="Purchase Request Line" duplicate="false">
+                    <sheet>
+                        <h1>
+                            <field name="request_id"/>
+                            <field name="request_state"/>
```
Comment (LoisRForgeFlow) [inline purchase_request/views/purchase_request_view.xml]: Again, this seems a weird place to have more fields apart from name, state would fit better in a status bar as in almost any other document/model.

**Review by LoisRForgeFlow — APPROVED** [review]
Some remarks on views, apart from that functionally looks good!

File: purchase_request/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    "version": "11.0.1.0.2",
+    "summary": "Use this module to have notification of requirements of "
+               "materials and/or external services and keep track of such "
+               "requirements.",
+    "category": "Purchase Management",
+    "depends": [
+        "purchase",
+        "product"
+    ],
+    "data": [
```
Comment (nikul-serpentcs) [inline purchase_request/__manifest__.py:17]: @hveficent IMO `data` sequence like [this](https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L29)

File: purchase_request/models/purchase_request.py [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+    @api.multi
+    @api.depends('state')
+    def _compute_is_editable(self):
+        for rec in self:
+            if rec.state in ('to_approve', 'approved', 'rejected', 'done'):
+                rec.is_editable = False
+            else:
+                rec.is_editable = True
+
+    name = fields.Char('Request Reference', size=32, required=True,
```
Comment (nikul-serpentcs) [inline purchase_request/models/purchase_request.py]: IMO remove `size` attribute everywhere, It's not needed.

File: purchase_request/models/purchase_request.py [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+            new_qty = qty + po_line.product_qty
+
+        qty = max(rl_qty, supplierinfo_min_qty, new_qty)
+
+        return qty
+
+    @api.multi
+    def unlink(self):
+        if self.mapped('purchase_lines'):
+            raise Warning(
```
Comment (nikul-serpentcs) [inline purchase_request/models/purchase_request.py]: IMO use `UserError

File: purchase_request/views/purchase_order_view.xml [hunk 7]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0"?>
+<!-- Copyright 2018 Eficent Business and IT Consulting Services S.L.
+     License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl-3.0) -->
+<odoo>
+    <data>
```
Comment (nikul-serpentcs) [inline purchase_request/views/purchase_order_view.xml]: Remove `<data>`

File: purchase_request/wizard/purchase_request_line_make_purchase_order.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+                    and line_company_id != company_id:
+                raise exceptions.Warning(
+                    _('You have to select lines '
+                      'from the same company.'))
+            else:
+                company_id = line_company_id
+
+            line_picking_type = line.request_id.picking_type_id or False
+            if not line_picking_type:
+                raise exceptions.Warning(
```
Comment (nikul-serpentcs) [inline purchase_request/wizard/purchase_request_line_make_purchase_order.py]: use `UserError`

File: purchase_request/wizard/purchase_request_line_make_purchase_order.py [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+                    _('You have to enter a Picking Type.'))
+            if picking_type is not False \
+                    and line_picking_type != picking_type:
+                raise exceptions.Warning(
```
Comment (nikul-serpentcs) [inline purchase_request/wizard/purchase_request_line_make_purchase_order.py]: same here

File: purchase_request/wizard/purchase_request_line_make_purchase_order_view.xml [hunk 10]
```diff
@@ -0,0 +1,65 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2018 Eficent Business and IT Consulting Services S.L.
+     License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl-3.0) -->
+<odoo>
+    <data>
```
Comment (nikul-serpentcs) [inline purchase_request/wizard/purchase_request_line_make_purchase_order_view.xml]: remove `<data>`

**Review by nikul-serpentcs — COMMENTED** [review]
Improve Code

**Comment by etobella** [comment]
Just a question, the functionality is similar to stock_request_purchase, or it has some differences?

**Comment by HviorForgeFlow** [comment]
They are alike, while in stock_request_purchase is an aiming solution to manage stockable products in a way much easier, in purchase_request you can request any kind of product, even services. So, it's a module that can be used inter-departmentally to demand  for requirements, for example, new coffe machine, but also for stock requirements.

**Comment by MiquelRForgeFlow** [comment]
@hveficent could you squash a bit?

**Comment by HviorForgeFlow** [comment]
Sure

**Review by MiquelRForgeFlow — APPROVED** [review]
LGTM

**Review by bjeficent — CHANGES_REQUESTED** [review]
When creating RFQ only vendors that are companies should be selectable.

supplier

The purchase orders should be filtered by the selected vendor.

po

**Comment by HviorForgeFlow** [comment]
Comments attended @bjeficent. Thanks for reviewing.

File: purchase_request/wizard/purchase_request_line_make_purchase_order_view.xml [hunk 11]
```diff
@@ -12,7 +12,8 @@
                  <separator string="Existing RFQ to update:"/>
                  <newline/>
                  <group>
-                    <field name="purchase_order_id"/>
+                    <field name="purchase_order_id"
+                        domain="[('partner_id', '=', supplier_id)]"/>
```
Comment (MiquelRForgeFlow) [inline purchase_request/wizard/purchase_request_line_make_purchase_order_view.xml:16]: if supplier_id is False, then what happens?

Comment (HviorForgeFlow) [reply purchase_request/wizard/purchase_request_line_make_purchase_order_view.xml:16]: you can't create a PO.

**Review by AaronHForgeFlow — APPROVED** [review]
:+1: Functional review

File: purchase_request/wizard/purchase_request_line_make_purchase_order_view.xml (hunk 11 above)
Comment (MiquelRForgeFlow) [reply purchase_request/wizard/purchase_request_line_make_purchase_order_view.xml:16]: then maybe the field `supplier_id` should be `required=True`

**Review by bjeficent — APPROVED** [review]
Tested in local :+1:

File: purchase_request/wizard/purchase_request_line_make_purchase_order.py [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+                                 related='line_id.product_id')
+    name = fields.Char(string='Description', required=True)
+    product_qty = fields.Float(string='Quantity to purchase',
+                               digits=dp.get_precision('Product UoS'))
+    product_uom_id = fields.Many2one('product.uom', string='UoM')
+    keep_description = fields.Boolean(string='Copy descriptions to new PO',
+                                      help='Set true if you want to keep the '
+                                           'descriptions provided in the '
+                                           'wizard in the new PO.',
+                                      default=False)
```
Comment (nikul-serpentcs) [inline purchase_request/wizard/purchase_request_line_make_purchase_order.py]: @hveficent `Boolean` always default False, so no need to `default=False`

File: purchase_request/wizard/purchase_request_line_make_purchase_order.py [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+            'product_id': product.id,
+            'product_uom': product.uom_po_id.id,
+            'price_unit': 0.0,
+            'product_qty': qty,
+            'account_analytic_id': item.line_id.analytic_account_id.id,
+            'purchase_request_lines': [(4, item.line_id.id)],
+            'date_planned': item.line_id.date_required,
+            'move_dest_ids': [(4, x.id) for x in item.line_id.move_dest_ids]
+        }
+        # if item.line_id.procurement_id:
```
Comment (nikul-serpentcs) [inline purchase_request/wizard/purchase_request_line_make_purchase_order.py]: remove Dead Code

File: purchase_request/wizard/purchase_request_line_make_purchase_order.py [hunk 14]
```diff
@@ -0,0 +1,306 @@
+# Copyright 2018 Eficent Business and IT Consulting Services S.L.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl-3.0).
+
+
+from odoo import api, fields, models, _
+import odoo.addons.decimal_precision as dp
+from odoo.exceptions import UserError, Warning
```
Comment (nikul-serpentcs) [inline purchase_request/wizard/purchase_request_line_make_purchase_order.py]: Remove `Warning`  Use only `UserError` everywhere

File: purchase_request/models/product_template.py [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    purchase_request = fields.Boolean(
+        string='Purchase Request',
+        help="Check this box to generate purchase request instead of "
+             "generating requests for quotation from procurement.",
+        default=False,
```
Comment (nikul-serpentcs) [inline purchase_request/models/product_template.py]: `Boolean` always  default False, So no need to `default=False`.

File: purchase_request/models/purchase_order.py [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+                request.message_post(body=message, subtype='mail.mt_comment')
+        return True
+
+    @api.multi
+    def _purchase_request_line_check(self):
+        for po in self:
+            for line in po.order_line:
+                for request_line in line.purchase_request_lines:
+                    if request_line.sudo().purchase_state == 'done':
+                        raise exceptions.Warning(
```
Comment (nikul-serpentcs) [inline purchase_request/models/purchase_order.py]: @hveficent use `UserError`

File: purchase_request/security/purchase_request.xml [hunk 17]
```diff
@@ -0,0 +1,111 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2018 Eficent Business and IT Consulting Services S.L.
+     License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl-3.0) -->
+<odoo>
+<data noupdate="0">
```
Comment (nikul-serpentcs) [inline purchase_request/security/purchase_request.xml]: if `noupdate="0"` then remove `data`

File: purchase_request/security/purchase_request.xml [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    </record>
+
+    <record id="group_purchase_request_manager" model="res.groups">
+        <field name="name">Purchase Request Manager</field>
+        <field name="implied_ids" eval="[(4, ref('purchase_request.group_purchase_request_user'))]"/>
+        <field name="category_id" ref="module_category_purchase_request"/>
+    </record>
+
+</data>
+<data noupdate="0">
```
Comment (nikul-serpentcs) [inline purchase_request/security/purchase_request.xml]: same here

File: purchase_request/views/purchase_request_view.xml [hunk 19]
```diff
@@ -0,0 +1,589 @@
+<?xml version="1.0"?>
+<!-- Copyright 2018 Eficent Business and IT Consulting Services S.L.
+     License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl-3.0) -->
+<odoo>
+
+    <record model="ir.ui.view" id="view_purchase_request_form">
+        <field name="name">purchase.request.form</field>
+        <field name="model">purchase.request</field>
+        <field name="arch" type="xml">
+            <form string="Purchase Request" version="7.0">
```
Comment (nikul-serpentcs) [inline purchase_request/views/purchase_request_view.xml]: remove `version`

File: purchase_request/views/purchase_request_view.xml [hunk 20]
```diff
@@ -0,0 +1,589 @@
...
+                                                       class="oe_inline"
+                                                       attrs="{'readonly': [('is_editable','=', False)]}"/>
+                                            </div>
+                                            <field name="analytic_account_id"
+                                                   groups="analytic.group_analytic_accounting"
+                                                   attrs="{'readonly': [('is_editable','=', False)]}"/>
+                                            <field name="date_required"
+                                                   attrs="{'readonly': [('is_editable','=', False)]}"/>
+                                            <field name="cancelled"/>
+                                            <!--<field name="procurement_id"/>-->
```
Comment (nikul-serpentcs) [inline purchase_request/views/purchase_request_view.xml]: remove Dead Code

File: purchase_request/wizard/purchase_request_line_make_purchase_order.py [hunk 21]
```diff
@@ -0,0 +1,306 @@
...
+    supplier_id = fields.Many2one('res.partner', string='Supplier',
+                                  required=True,
+                                  domain=[('supplier', '=', True),
+                                          ('is_company', '=', True)])
+    item_ids = fields.One2many(
+        'purchase.request.line.make.purchase.order.item',
+        'wiz_id', string='Items')
+    purchase_order_id = fields.Many2one('purchase.order',
+                                        string='Purchase Order',
+                                        required=False,
```
Comment (nikul-serpentcs) [inline purchase_request/wizard/purchase_request_line_make_purchase_order.py]: no need to `required=False`

**Review by nikul-serpentcs — CHANGES_REQUESTED** [review]
Improve Code

**Comment by HviorForgeFlow** [comment]
Thanks @nikul-serpentcs for re-reviewing! Comments attended ;)

**Review by nikul-serpentcs — APPROVED** [review]
Code Review LGTM :+1:

**Pull request merged**